From afd7606ca42a2586b8823c7bd1a4a7cfd2476e3b Mon Sep 17 00:00:00 2001
From: Steven J. Magnani <steve@digidescorp.com>
Date: Wed, 09 Jun 2010 14:02:21 +0000
Subject: malloc-simple: Make calloc() return zeroed memory

The 0.9.31 release included a change to malloc-simple to request
uninitialized memory from noMMU kernels. Unfortunately, the corresponding
calloc() code assumed that memory returned by malloc() was already zeroed,
which leads to all kinds of nastiness.

Signed-off-by: Steven J. Magnani <steve@digidescorp.com>
Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
---
diff --git a/libc/stdlib/malloc-simple/alloc.c b/libc/stdlib/malloc-simple/alloc.c
index 51da14a..914c89d 100644
--- libc/stdlib/malloc-simple/alloc.c
+++ libc/stdlib/malloc-simple/alloc.c
@@ -60,11 +60,10 @@ void * calloc(size_t nmemb, size_t lsize)
 		__set_errno(ENOMEM);
 		return NULL;
 	}
-	result=malloc(size);
-#if 0
-	/* Standard unix mmap using /dev/zero clears memory so calloc
-	 * doesn't need to actually zero anything....
-	 */
+	result = malloc(size);
+
+#ifndef __ARCH_USE_MMU__
+	/* mmap'd with MAP_UNINITIALIZE, we have to blank memory ourselves */
 	if (result != NULL) {
 		memset(result, 0, size);
 	}
